framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -285,104 +283,120 @@ protected boolean areEqualInHierarchy(
   }
 
   /**
-   * A declared type is considered a supertype of another declared type only if all of the type
-   * arguments of the declared type "contain" the corresponding type arguments of the subtype.
-   * Containment is described in <a
+   * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+   * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}.
+   *
+   * <p>Containment is described in <a
    * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.5.1">JLS section
    * 4.5.1 "Type Arguments of Parameterized Types"</a>.
    *
-   * @param inside a type argument of the "subtype"
-   * @param outside a type argument of the "supertype"
+   * <p>As described in <a
+   * href=https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.10.2>JLS section
+   * 4.10.2 Subtyping among Class and Interface Types</a>, a declared type is considered a supertype
+   * of another declared type only if all of the type arguments of the declared type "contain" the
+   * corresponding type arguments of the subtype.
+   *
+   * @param inside a possibly-contained type
+   * @param outside a possibly-containing type
    * @param canBeCovariant whether or not type arguments are allowed to be covariant
    * @return true if inside is contained by outside, or if canBeCovariant == true and {@code inside
    *     <: outside}
    */
   protected boolean isContainedBy(
-      final AnnotatedTypeMirror inside, final AnnotatedTypeMirror outside, boolean canBeCovariant) {
+      AnnotatedTypeMirror inside, AnnotatedTypeMirror outside, boolean canBeCovariant) {
 
-    if (ignoreUninferredTypeArgument(inside) || ignoreUninferredTypeArgument(outside)) {
+    if (shouldIgnoreUninferredTypeArgs(inside) || shouldIgnoreUninferredTypeArgs(outside)) {
       areEqualVisitHistory.put(inside, outside, currentTop, true);
       return true;
     }
 
-    if (outside.getKind() != TypeKind.WILDCARD
-        && !TypesUtils.isCaptured(outside.getUnderlyingType())) {
-      if (canBeCovariant) {
-        return isSubtype(inside, outside, currentTop);
+    if (outside.getKind() == TypeKind.WILDCARD) {
+      Boolean previousResult = areEqualVisitHistory.get(inside, outside, currentTop);
```

Comment:
I usually expect to see a check of a cache at the beginning of the routine.  Should it be moved there, or is there a reason that it needs to appear after other logic?